README.md
# PlugLimit

[![Hex Version](https://img.shields.io/hexpm/v/plug_limit)](https://hex.pm/packages/plug_limit)
[![Hex Docs](https://img.shields.io/badge/hex-docs-lightgreen)](https://hexdocs.pm/plug_limit)

Rate limiting Plug module based on Redis Lua scripting.

## Summary
PlugLimit is using Redis Lua scripting to provide rate-limiting functionality for web applications
based on a Plug library. PlugLimit has a modular architecture: users can use their own Redis Lua
scripts implementing custom rate limiting algorithms.

Salient Redis Lua scripting feature is a race conditions resiliency which makes this library
a recommended solution for distributed systems (e.g.: Phoenix servers behind round robin load balancer).

PlugLimit provides two built-in rate limiters: `PlugLimit.FixedWindow` and `PlugLimit.TokenBucket`.

## Installation
Add `PlugLimit` library to your application dependencies:
```elixir
def deps do
  [
    {:plug_limit, "~> 0.1.0"}
  ]
end
```

## Usage
PlugLimit is Redis client agnostic. In a first step you must define Elixir function executing Redis
commands, depending on Redis client of your choice:
```elixir
# config/config.exs
config :plug_limit,
  enabled?: true,
  cmd: {MyApp.Redis, :command, []}
```

`MyApp.Redis.command/2` function must accept Redis command as a first argument and static MFA tuple
`arg` as a second.
In most cases `:cmd` function will be a `Redix.command/3` or `:eredis.q/2,3` wrapper.
Example naive [Redix](https://hex.pm/packages/redix) driver wrapper:
```elixir
#lib/my_app/redis.ex
def command(command, opts \\ [timeout: 500]) do
  {:ok, pid} = Redix.start_link()
  Redix.command(pid, command, opts)
  :ok = Redix.stop(pid)
end
```

PlugLimit is tested with both [Redix](https://hex.pm/packages/redix) and
[eredis](https://hex.pm/packages/eredis) Redis clients.

Phoenix Framework endpoint can be protected with fixed window rate-limiter by placing a
`PlugLimit.FixedWindow` plug call in the request processing pipeline:
```elixir
#lib/my_app_web/router.ex
pipeline :high_cost_pipeline do
  plug(PlugLimit.FixedWindow,
    limit: 10,
    ttl: 60,
    key: {MyApp.RateLimiter, :user_key, [:high_cost_pipeline]}
  )
  # remaining pipeline plugs...
end
```

Rate limits for `:high_cost_pipeline` pipeline will be evaluated with Redis Lua script fixed window
algorithm. Client identified with `:key` will be allowed to issue 10 requests in 60 seconds time
window.

MFA tuple defined with `:key` option specifies user function which should provide Redis key
used to uniquely identify given rate-limiter bucket. Redis rate-limiter key name should be derived
from `Plug.Conn.t()` connection struct parameters.
Example function to create Redis key name for rate-limiter throttling requests for a given user
identified by a connection assigned `:user_id`:
```elixir
#lib/my_app/rate_limiter.ex
def user_key(%Plug.Conn{assigns: %{user_id: user_id}}, prefix),
   do: {:ok, ["#{prefix}:#{user_id}"]}
```

Please refer to `PlugLimit` module documentation for detailed library configuration description and
"Redis Lua script rate limiters" in LIMITERS.md file for Redis Lua scripts implementation
guidelines.

## TODO
- [ ] Add leaky bucket rate limiting algorithm implementation.


CHANGELOG.md
## v0.1.0 [2022-07-19]

Available on hex.pm.


lib/test.ex
defmodule PlugLimit.Test do
  @moduledoc """
  Conveniences for testing endpoints protected with PlugLimit rate limiters.

  Unit testing endpoints protected with rate limiters might provide unexpected results
  because after exceeding requests limit during consecutive tests given endpoint will respond
  with `429` status code.

  `PlugLimit.Test` can help to avoid unexpected `429` responses by adding following code to your
  test cases:
  ```elixir
  use PlugLimit.Test, redis_flushdb?: true
  ```
  Using this module with `:redis_flushdb?` set to `true` will flush Redis database with
  `redis_flushdb/1` function executed inside `ExUnit.Callbacks.setup/1`.
  Required by `redis_flushdb/1` Redis command function is taken from `:plug_limit` `:cmd` key
  defined for testing environment, for example:
  ```elixir
  # config/test.exs
  config :plug_limit, cmd: {MyApp.Redis, :command, []}
  ```
  Please refer to `PlugLimit` module documentation for detailed `:cmd` setting description.

  `PlugLimit.Test` has one configuration setting `:redis_flushdb?`, set to `false` by default.
  When using `:redis_flushdb?` you might consider setting separate Redis database
  (e.g.: `redis://localhost:6379/15`) for PlugLimit to avoid interference with other parts of your
  application during testing.

  Alternatively to `:redis_flushdb?` you can use `redis_del_keys/2`.
  """

  use ExUnit.CaseTemplate

  using opts do
    quote bind_quoted: [opts: opts] do
      @redis_flushdb Keyword.get(opts, :redis_flushdb?, false)

      setup do
        if @redis_flushdb,
          do: :plug_limit |> Application.fetch_env!(:cmd) |> PlugLimit.Test.redis_flushdb(),
          else: :ok
      end
    end
  end

  @doc """
  Gets `x-ratelimit-remaining` header from `conn` and parses it to integer.

  Returns `x-ratelimit-remaining` header as an integer or string message if invalid or missing header.

  Example:
  ```elixir

  test "x-ratelimit-remaining is correct", %{conn: conn} do
    conn = get(conn, "/")
    assert PlugLimit.Test.get_remaining(conn) == 60
  end
  ```

  ```elixir
  iex> PlugLimit.Test.get_remaining(conn_invalid_header)
  "Invalid, non-standard or missing x-ratelimit-remaining header."
  ```
  """
  @spec get_remaining(conn :: Plug.Conn.t()) :: integer() | String.t()
  def get_remaining(conn), do: hdr_to_integer(conn, "x-ratelimit-remaining")

  @doc """
  Same as `get_remaining/1` but for `x-ratelimit-reset` header.
  """
  @spec get_reset(conn :: Plug.Conn.t()) :: integer()
  def get_reset(conn), do: hdr_to_integer(conn, "x-ratelimit-reset")

  @doc """
  Checks for IETF recommended `x-ratelimit-*` http response headers in `conn`.

  Checks if basic rate limiting http response headers complying with
  ["RateLimit Fields for HTTP"](https://datatracker.ietf.org/doc/draft-ietf-httpapi-ratelimit-headers/)
  IETF specification are added to `Plug.Conn.t()`.

  Required rate limiting http response headers:
  * `x-ratelimit-limit`,
  * `x-ratelimit-remaining`,
  * `x-ratelimit-reset`.

  Returns `true` if above headers are present and `false` otherwise.

  Example:
  ```elixir
  test "it is protected by rate limiter", %{conn: conn} do
    conn = get(conn, "/")
    assert PlugLimit.Test.headers_exist?(conn)
  end
  ```
  """
  @spec headers_exist?(conn :: Plug.Conn.t()) :: boolean()
  def headers_exist?(conn) do
    ["x-ratelimit-limit", "x-ratelimit-remaining", "x-ratelimit-reset"]
    |> Enum.map(fn hdr ->
      conn
      |> Plug.Conn.get_resp_header(hdr)
      |> Enum.empty?()
      |> Kernel.not()
    end)
    |> Enum.reduce(true, fn v, acc -> acc && v end)
  end

  @redis_del_keys_script """
  return redis.call('DEL', 'default:not:existing:key:*', unpack(redis.call('KEYS', ARGV[1])))
  """

  @doc """
  Deletes all keys with the names matching pattern given by `key` from the Redis DB selected with
  `{m, f, a}` Redis command function.

  `{m, f, a}` usually should be consistent with a value given in `:plug_limit` `:cmd` configuration key.
  Returns `:ok` on success or `any()` on error.

  Function should not be used to delete large numbers of Redis keys.

  Example:
  ```elixir
  setup do
    :ok = PlugLimit.Test.redis_del_keys({MyApp.Redis, :command, []}, "my_key:*")
  end
  ```
  """
  @spec redis_del_keys({m :: module(), f :: atom(), [a :: any()]}, key :: String.t()) ::
          :ok | any()
  def redis_del_keys({m, f, a} = _redis_command_function, key) do
    case apply(m, f, [["EVAL", @redis_del_keys_script, "0", [key]]] ++ a) do
      {:ok, _key_count} -> :ok
      err -> err
    end
  end

  @doc """
  Deletes all keys from the Redis DB selected with `{m, f, a}` Redis command function.

  `{m, f, a}` usually should be consistent with a value given in `:plug_limit` `:cmd` configuration key.
  Returns `:ok` on success or `any()` on error.

  Redis command which is used to delete keys: [`FLUSHDB SYNC`](https://redis.io/commands/flushdb/).

  Example:
  ```elixir
  iex> PlugLimit.Test.redis_flushdb({MyApp.Redis, :command, []})
  :ok
  ```
  """
  @spec redis_flushdb({m :: module(), f :: atom(), [a :: any()]}) :: :ok | any()
  def redis_flushdb({m, f, a} = _redis_command_function) do
    case apply(m, f, [["FLUSHDB"]] ++ a) do
      {:ok, "OK"} -> :ok
      err -> err
    end
  end

  defp hdr_to_integer(conn, hdr) do
    with [hdr] <- Plug.Conn.get_resp_header(conn, hdr),
         {hdr, ""} <- Integer.parse(hdr) do
      hdr
    else
      _ -> "Invalid, non-standard or missing #{hdr} header."
    end
  end
end


test/test_test.exs
defmodule PlugLimit.TestTest do
  use ExUnit.Case
  use Plug.Test

  setup do
    conn = PlugLimit.Test.Helpers.build_conn()
    %{conn: conn}
  end

  describe "get_remaining/1" do
    test "valid header", %{conn: conn} do
      conn = Plug.Conn.put_resp_header(conn, "x-ratelimit-remaining", "123456")
      assert PlugLimit.Test.get_remaining(conn) == 123_456
    end

    test "invalid header", %{conn: conn} do
      conn = Plug.Conn.put_resp_header(conn, "x-ratelimit-remaining", "abc")

      assert PlugLimit.Test.get_remaining(conn) ==
               "Invalid, non-standard or missing x-ratelimit-remaining header."
    end

    test "missing header", %{conn: conn} do
      assert PlugLimit.Test.get_remaining(conn) ==
               "Invalid, non-standard or missing x-ratelimit-remaining header."
    end
  end

  describe "get_reset/1" do
    test "valid header", %{conn: conn} do
      conn = Plug.Conn.put_resp_header(conn, "x-ratelimit-reset", "123456")
      assert PlugLimit.Test.get_reset(conn) == 123_456
    end

    test "invalid header", %{conn: conn} do
      conn = Plug.Conn.put_resp_header(conn, "x-ratelimit-reset", "abc")

      assert PlugLimit.Test.get_reset(conn) ==
               "Invalid, non-standard or missing x-ratelimit-reset header."
    end

    test "missing header", %{conn: conn} do
      assert PlugLimit.Test.get_reset(conn) ==
               "Invalid, non-standard or missing x-ratelimit-reset header."
    end
  end

  describe "headers_exist?/1" do
    test "headers exist", %{conn: conn} do
      conn =
        conn
        |> Plug.Conn.put_resp_header("x-ratelimit-limit", "123")
        |> Plug.Conn.put_resp_header("x-ratelimit-remaining", "456")
        |> Plug.Conn.put_resp_header("x-ratelimit-reset", "789")

      assert PlugLimit.Test.headers_exist?(conn)
    end

    test "headers do not exist", %{conn: conn} do
      refute PlugLimit.Test.headers_exist?(conn)
    end
  end
end


test/test_helper.exs
ExUnit.start()

{:ok, pid} = Redix.start_link()
:ok = :persistent_term.put({:redis_cli, :redix}, pid)

{:ok, pid} = :eredis.start_link()
:ok = :persistent_term.put({:redis_cli, :eredis}, pid)

defmodule PlugLimit.Test.RedixCli do
  def command(command, opts \\ [timeout: 5000]) do
    {:redis_cli, :redix}
    |> :persistent_term.get()
    |> Redix.command(command, opts)
  end
end

defmodule PlugLimit.Test.EredisCli do
  def command(command, timeout \\ 5000) do
    {:redis_cli, :eredis}
    |> :persistent_term.get()
    |> :eredis.q(command, timeout)
  end
end

defmodule PlugLimit.Test.ErrorCli do
  def command(_command, _opts), do: {:error, %{reason: :closed}}
end

defmodule PlugLimit.Test.Helpers do
  def build_conn(), do: :get |> Plug.Test.conn("/") |> Plug.Conn.assign(:user_id, 123)

  def user_id_key(%Plug.Conn{assigns: %{user_id: user_id}}, prefix),
    do: {:ok, ["#{prefix}:#{user_id}"]}

  def put_response_echo(%Plug.Conn{} = conn, conf, eval_result, args),
    do: {conn, conf, eval_result, args}
end


mix.exs
defmodule PlugLimit.MixProject do
  use Mix.Project

  @source_url "https://github.com/Recruitee/plug_limit"
  @version "0.1.0"

  def project do
    [
      app: :plug_limit,
      version: "0.1.0",
      elixir: "~> 1.10",
      start_permanent: Mix.env() == :prod,
      deps: deps(),
      docs: docs(),
      package: package(),
      name: "PlugLimit",
      description: "Rate limiting Plug module based on Redis Lua scripting"
    ]
  end

  def application do
    [extra_applications: [:logger]]
  end

  defp deps do
    [
      {:plug, "~> 1.10"},
      {:ex_doc, "~> 0.28.0", only: :dev, runtime: false},
      {:eredis, "~> 1.4", only: :test},
      {:redix, "~> 1.1", only: :test}
    ]
  end

  defp package do
    [
      files: ~w(lib lua .formatter.exs mix.exs CHANGELOG.md README.md LIMITERS.md),
      maintainers: ["Recruitee", "Andrzej Magdziarz"],
      licenses: ["Apache-2.0"],
      links: %{"GitHub" => @source_url}
    ]
  end

  defp docs do
    [
      extras: ["README.md", "LIMITERS.md"],
      main: "readme",
      source_url: @source_url,
      source_ref: "v#{@version}",
      skip_undefined_reference_warnings_on: ["CHANGELOG.md"]
    ]
  end
end


lib/plug_limit.ex
defmodule PlugLimit do
  @moduledoc """
  Rate limiting Plug module based on Redis Lua scripting.

  ## Architecture
  `PlugLimit` is a `Plug` behaviour module implementation using
  [Redis Lua scripting](https://redis.io/docs/manual/programmability/eval-intro/) to provide
  rate limiting functionality.
  Module plugs should implement two callback functions: `c:Plug.init/1` and `c:Plug.call/2`.
  `PlugLimit` `init/1` function is responsible for building plug configuration, please refer to the
  [Configuration](#module-configuration) section below for details.

  `PlugLimit` `call/2` function has following responsibilities:
  1. Establish if given request should be allowed or denied in accordance with selected rate
     limiting strategy.
  2. Determine values of the rate limiting http headers.
  3. Set rate limiting http headers for the response.
  4. Halt request processing pipeline if rate limit was exceeded and send response with
     appropriate status code, usually `429 - Too Many Requests`.

  Tasks 1 and 2 are performed by evaluation of the Redis Lua script implementing rate limiting
  algorithm.
  Tasks 3 and 4 are executed by built-in `put_response/4` function or user provided equivalent
  callback function leveraging Redis Lua script evaluation results.

  PlugLimit Redis Lua scripts are loaded to the Redis scripts cache on the first `c:Plug.call/2`
  callback invocation by using Redis [`SCRIPT LOAD`](https://redis.io/commands/script-load/)
  command.
  Generated by `SCRIPT LOAD` SHA1 script hash is cached locally as a `:persistent_term` key.
  Lua script SHA1 hash is later retrieved from `:persistent_term` local cache and used with
  Redis [`EVALSHA`](https://redis.io/commands/evalsha/) command on subsequent rate-limiter
  Lua script evaluations.
  Implemented SHA1 Lua script caching mechanism is resilient to Redis script cache resets by Redis
  instance reboots or Redis [`SCRIPT FLUSH`](https://redis.io/commands/script-flush/) command use.

  Redis Lua script evaluation is an atomic operation resilient to the race conditions in distributed
  environments.

  In normal circumstances latency introduced by Redis `EVALSHA` command should be close to a single
  Redis request/response round trip time, usually less than ~1 ms.

  Redis Lua script execution blocks single-threaded Redis server, so it is advised to use a separate
  standalone Redis instance for PlugLimit rate-limiters, especially when using custom untested
  Lua script implementations.
  Data stored by Redis rate-limiters in most cases can be considered as strongly interim so
  using highly available Redis Sentinel instances for rate-limiters data might be unnecessary.
  For high traffic volume cases, sharding can be easily achieved by distributing independent Phoenix
  router pipelines or scopes rate-limiters between dedicated Redis instances.

  ## Usage
  PlugLimit in most basic use case requires configuration of a function which will execute Redis
  commands:
  ```elixir
  # config/config.exs
  config :plug_limit,
    enabled?: true,
    cmd: {MyApp.Redis, :command, []}
  ```

  When working with Phoenix Framework `PlugLimit` plug call can be placed at the endpoint, router
  or controller depending on requirements.
  Example of minimal `PlugLimit` call used in `:high_cost_pipeline` router pipeline:
  ```elixir
  #lib/my_app_web/router.ex
  pipeline :high_cost_pipeline do
    plug(PlugLimit, opts: [10, 60], key: {MyApp.RateLimiter, :user_id_key, [:high_cost_pipeline]})
    # remaining pipeline plugs...
  end
  ```

  Example above will evaluate request rate-limiting parameters using built-in default Lua script
  implementing fixed window algorithm with rate limiter algorithm options given in the `:opts` list.
  First list element specifies request limit, set to 10 here, second item specifies limiting time
  window length in seconds, set to 60.
  Redis rate-limiter bucket name will be evaluated with function defined by the `:key` MFA tuple.
  Example Redis key bucket name function result for `user_id=12345`:
  `{:ok, ["high_cost_pipeline_limiter:12345"]}`.

  Example `PlugLimit` configuration for built-in token bucket rate-limiter:
  ```elixir
  #lib/my_app_web/router.ex
  pipeline :high_cost_pipeline do
    plug(PlugLimit,
      limiter: :token_bucket,
      opts: [20, 600, 5],
      key: {MyApp.RateLimiter, :user_id_key, ["high_cost_pipeline"]}
    )
    # remaining pipeline plugs...
  end
  ```

  Configuration options details are described in the [Configuration](#module-configuration)
  section below.

  Instead of using generic `PlugLimit` module you can use provided convenience wrappers:
  `PlugLimit.FixedWindow` or `PlugLimit.TokenBucket`.

  Built-in rate-limiting algorithms are described in the "Redis Lua script rate limiters"
  LIMITERS.md file.

  Unit testing for user applications using PlugLimit library is described in "Unit testing"
  TESTING.md file.

  ## Configuration
  PlugLimit configuration is built from following sources:
  * global `:plug_limit` configuration parameters from application configuration file, usually
    `config/*.exs`,
  * parameters overwriting global configuration passed as arguments to the PlugLimit plug
    call in the application router or controller,
  * hard-coded default values.

  `PlugLimit` is using a concept of rate-limiters to organize individual limiters configurations.
  Rate-limiters configurations are declared in application configuration file using `:limiters`
  key. Each rate-limiter must be associated with a valid Lua script. Lua scripts are configured
  with `:luascripts` key.

  Full example configuration defining `:custom_limiter` limiter using `:custom_bucket` Lua script:
  ```elixir
  # config/config.exs
  config :plug_limit,
    limiters: [
      custom_limiter: %{
        cmd: {MyApp.Redis, :command, ["redis://10.10.10.2:6379/"]},
        key: {MyApp.RateLimiter, :user_id_key, []},
        log_level: :error,
        luascript: :custom_bucket,
        response: {MyApp.RateLimiter, :respond, []}
      }
    ],
    luascripts: [
      custom_bucket: %{
        script: {File, :read, ["./lua/custom_bucket.lua"]},
        headers: [
          "x-ratelimit-limit",
          "x-ratelimit-reset",
          "x-ratelimit-remaining",
          "x-acme-custom-header"
        ]
      }
    ]
  ```
  Above defined `:custom_limiter` should be referred as follows:
  ```elixir
  plug(PlugLimit, limiter: :custom_limiter, opts: [20, 600, 5])
  ```

  ### Application configuration options
  Primary part of `PlugLimit` configuration is located in the application configuration file,
  usually `config/config.exs`, under `:plug_limit` key. Some of the options defined in application
  configuration can be overwritten with `plug(PlugLimit, [options...])` call.

  Available configuration options:
  * `:enabled?` - when set to Boolean `false` or string `"false"` `PlugLimit` is disabled and
    `plug(PlugLimit, opts: [...])` call immediately returns unmodified `conn` struct.
    To enable `PlugLimit`, `:enabled?` key must be set to Boolean `true` or string `"true"`.
    Default: `false`.
  * `:cmd` - MFA tuple pointing at the user defined two arity function executing Redis commands.
    As a first argument function will receive a Redis command as a list, for example:
    `["SET", "mykey", "foo"]`.
    As a second parameter, static argument defined in the MFA tuple will be passed.
    Redis command function should return `{:ok, redis_response}` on success and `{:error, reason}`
    on error.
    When using [Redix](https://hex.pm/packages/redix) library as a client, `:cmd` command should be
    a `Redix.command/3` wrapper.
    When using [eredis](https://hex.pm/packages/eredis) library, wrapper for the `:eredis.q/2,3`
    should be implemented.
    Redis command function defined here will be used as a default function for limiters that
    do not have their own `:cmd` specified.
    Optional if each limiter has its own `:cmd` defined, required otherwise.
  * `:log_level` - specifies log level for rate-limiters that do not set their own `:log_level`.
    Only library errors are logged - with `put_response/4` function.
    Boolean value `false` disables logging. Please refer to the `Logger` documentation
    for valid log levels. Default: `:error`.
  * `:response` - MFA tuple pointing at the user defined 4 arity function providing request
    response depending on rate-limiter Lua script evaluation results.
    Please refer to the built-in response function `put_response/4` description for details.
    Can be overwritten for individual limiters.
    Default: `put_response/4`.
  * `:limiters` - keyword list with user provided rate-limiters. See below for details. Optional.
  * `:luascripts` - keyword list defining Lua scripts for rate-limiters defined with `:limiters`.
    See below for details. Optional.

  PlugLimit `:enabled?` option is the only option evaluated at run-time with `c:Plug.call/2`,
  so function like `System.get_env/2` can be used here.
  All other configuration options are initialized with `c:Plug.init/1`, which usually takes place at
  compile time for production or release environments and run-time for testing and development.
  In a production environment, `PlugLimit` `:enabled` might be controlled using environmental
  variable, for example:
  ```elixir
  # config/config.exs
  config :plug_limit, cmd: {MyApp.Redis, :command, []}

  # config/releases.exs
  config :plug_limit, enabled?: System.get_env("PLUG_LIMIT_ENABLED", "false")

  # config/dev.exs
  config :plug_limit, enabled?: true

  # config/test.exs
  config :plug_limit, enabled?: true
  ```

  Custom user rate-limiters are configured as a `:limiters` keyword list.
  Rate-limiters are declared as maps with following keys:
  * `:cmd` - overwrites `:cmd` global key for a given rate-limiter. Optional.
  * `:key` - MFA tuple pointing at the user defined two arity function providing Redis keys names
    that will be passed later to the Redis Lua script.
    Function receives request `Plug.Conn.t()` struct as a first argument and static argument from
    the MFA tuple as a second argument.
    Function should return `{:ok, [key :: String.t()]}` when successful and `{:error, reason}`
    on error. Function should return especially name of the key that Redis Lua script will use to
    create a unique bucket for a given rate-limiter and requests group.
    Please refer to "Redis Lua script rate limiters" LIMITERS.md file for further discussion.
    Example `:key` function implementation:
    ```elixir
    def user_key(%Plug.Conn{assigns: %{user_id: user_id}}, prefix),
      do: {:ok, [to_string(prefix) <> ":" <> to_string(user_id)]}

    def user_key(_conn, _prefix), do: {:error, "Missing user_id"}
    ```
    Redis keys names should follow
    [Redis keys naming conventions](https://redis.io/docs/manual/data-types/data-types-tutorial/#keys).
    `:key` value can be overwritten in a plug call configuration.
    Optional if `:key` is specified for each `PlugLimit` plug call, required otherwise.
  * `:log_level` - overwrites global `:log_level` for a given rate-limiter. Optional.
  * `:luascript` - atom defining Lua script for a given rate-limiter. Lua scripts are defined with
    `:luascripts` keyword list, see below for details. Required.
  * `:response` - overwrites global `:response` for a given rate-limiter. Optional.

  PlugLimit provides two built-in rate-limiters: `:fixed_window` and `:token_bucket`, please refer
  to "Redis Lua script rate limiters" LIMITERS.md file for details.

  Each rate-limiter is associated with Redis Lua script checking if given request should
  be allowed or denied and evaluating rate limiting http headers. Redis Lua scripts are configured
  as `:luascripts` keyword list. Each script is declared as a map with following keys:
  * `:script` - MFA tuple pointing at one arity function returning
    `{:ok, limiter_script :: String.t()}` on success and `{:error, reason}` on error.
    Function receives as an argument static argument from the MFA tuple.
    Example implementations:
    ```elixir
    def get_lua_script_by_path(path), do: File.read(path)

    def my_lua_script(_arg), do: {:ok, "-- Lua script body"}
    ```
    Required.
  * `:headers` - list of rate limiting headers keys to be used with headers values returned by
    given Redis Lua script to build request response http headers.
    Example headers list:
    ```elixir
    headers: [
      "x-ratelimit-limit",
      "x-ratelimit-reset",
      "x-ratelimit-remaining"
    ]
    ```
    Please refer to "Redis Lua script rate limiters" LIMITERS.md file for more detailed discussion
    on rate limiting headers. Required.

  ### Plug call configuration options
  * `:key` - MFA tuple, overwrites value given in limiter's application configuration.
    Required if not provided in the limiter configuration.
  * `:limiter` - atom selecting rate limiter. List of built-in limiters is provided in LIMITERS.md
    file. Default: `:fixed_window`.
  * `:opts` - list with rate limiting options like requests limit, time window length or burst rate.
    List is passed as an argument to the Redis Lua script, see LIMITERS.md for built-in limiters
    options. Required.
  """

  @behaviour Plug
  import Plug.Conn
  require Logger

  @default_limiter_id :fixed_window

  @default_limiters [
    fixed_window: %{
      luascript: :fixed_window
    },
    token_bucket: %{
      luascript: :token_bucket
    }
  ]

  @default_luascripts [
    fixed_window: %{
      script: {__MODULE__, :get_script, [:fixed_window]},
      headers: [
        "x-ratelimit-limit",
        "x-ratelimit-reset",
        "x-ratelimit-remaining"
      ]
    },
    token_bucket: %{
      script: {__MODULE__, :get_script, [:token_bucket]},
      headers: [
        "x-ratelimit-limit",
        "x-ratelimit-reset",
        "x-ratelimit-remaining",
        "retry-after"
      ]
    }
  ]

  @default_log_level :error

  @default_response {__MODULE__, :put_response, []}

  @default_script_fixed_window File.read("./lua/fixed_window.lua")
  @default_script_token_bucket File.read("./lua/token_bucket.lua")

  @limit_status 429
  @limit_body "Too Many Requests"

  @no_script_msg ["NOSCRIPT No matching script. Please use EVAL."]

  defstruct [
    :cmd,
    :headers,
    :key,
    :log_level,
    :opts,
    :response,
    :script,
    :script_id
  ]

  @type t() :: %__MODULE__{
          cmd: mfa(),
          headers: [String.t()],
          key: mfa(),
          log_level: log_level(),
          opts: list(),
          response: mfa(),
          script: mfa(),
          script_id: atom()
        }
  @type limiters :: [
          {limiter_id :: atom, limiter :: limiter()}
        ]
  @type limiter :: %{
          cmd: mfa(),
          key: mfa(),
          log_level: log_level(),
          luascript: atom(),
          response: mfa()
        }
  @type luascripts :: [
          {luascript_id :: atom(), luascript :: luascript()}
        ]
  @type luascript :: %{
          headers: [String.t()],
          opts: list(),
          script: mfa()
        }
  @type log_level :: false | Logger.level()

  @type eval_result :: {:ok, list()} | {:error, any()} | any()

  @impl true
  @doc false
  @spec init(opts :: Plug.opts()) :: PlugLimit.t()
  def init(opts) do
    limiters = Application.get_env(:plug_limit, :limiters, @default_limiters)
    luascripts = Application.get_env(:plug_limit, :luascripts, @default_luascripts)

    limiter_id = Keyword.get(opts, :limiter, @default_limiter_id)
    limiter = Keyword.get(limiters, limiter_id) || Keyword.fetch!(@default_limiters, limiter_id)
    script_id = Map.fetch!(limiter, :luascript)
    key = Keyword.get(opts, :key) || Map.fetch!(limiter, :key)

    luascript =
      Keyword.get(luascripts, script_id) || Keyword.fetch!(@default_luascripts, script_id)

    script = Map.fetch!(luascript, :script)
    headers = Map.fetch!(luascript, :headers)
    l_opts = Keyword.fetch!(opts, :opts)

    cmd = Map.get(limiter, :cmd) || Application.fetch_env!(:plug_limit, :cmd)

    log_level =
      Map.get(limiter, :log_level) ||
        Application.get_env(:plug_limit, :log_level, @default_log_level)

    response =
      Map.get(limiter, :response) ||
        Application.get_env(:plug_limit, :response, @default_response)

    opts = %{
      cmd: cmd,
      headers: headers,
      key: key,
      log_level: log_level,
      opts: l_opts,
      response: response,
      script: script,
      script_id: script_id
    }

    struct!(__MODULE__, opts)
  end

  @impl true
  @doc false
  @spec call(conn :: Plug.Conn.t(), conf :: Plug.opts()) :: Plug.Conn.t()
  def call(conn, conf) do
    case Application.get_env(:plug_limit, :enabled?, false) do
      res when res in ["true", true] ->
        eval_result = eval_limit(conn, conf)
        apply_response(conn, conf, eval_result)

      res when res in ["false", false] ->
        conn
    end
  end

  @doc """
  Puts new rate limiting http response headers in the connection and halts the Plug pipeline if
  rate limit was exceeded.

  `put_response/4` is a default `PlugLimit` function preparing an http response accordingly
  with Redis Lua script evaluation results.
  Custom response function can be selected by setting `:response` global or given limiter
  configuration keys.

  Function accepts following arguments:
  1. `Plug.Conn.t()` connection.
  2. Rate-limiter configuration as a `PlugLimit.t()` struct.
  3. Redis Lua script evaluation result as a `PlugLimit.eval_result()` type.
  4. Static argument given in the `:response` MFA tuple.

  Function returns `Plug.Conn.t()` struct with rate limiting headers.
  If rate limit is exceeded function halts Plug pipeline and sends response with `429` status code
  and plain-text body `"Too Many Requests"`.
  If Redis Lua script evaluation or any other rate-limiting processing function fails,
  `put_response/4` function will log resulting error with Logger level set by `:log_level`
  configuration setting and return unmodified connection struct.

  Custom response functions and custom Redis Lua scripts are described in more details in
  "Redis Lua script rate limiters" LIMITERS.md file.
  """
  @spec put_response(
          conn :: Plug.Conn.t(),
          conf :: PlugLimit.t(),
          eval_result :: eval_result(),
          args :: any()
        ) :: Plug.Conn.t()
  def put_response(conn, %__MODULE__{} = conf, eval_result, _args) do
    case eval_result do
      {:ok, [action, headers | _other]} ->
        conn = put_headers(conn, headers, conf.headers)

        if action == "allow" do
          conn
        else
          conn
          |> put_resp_content_type("text/plain")
          |> send_resp(@limit_status, @limit_body)
          |> halt()
        end

      err ->
        log_level = Map.fetch!(conf, :log_level)
        if log_level, do: Logger.log(log_level, "[#{__MODULE__}] #{inspect(err)}")
        conn
    end
  end

  @doc false
  def get_script(:fixed_window), do: @default_script_fixed_window
  def get_script(:token_bucket), do: @default_script_token_bucket

  @spec eval_limit(Plug.Conn.t(), PlugLimit.t()) :: eval_result()
  defp eval_limit(conn, %__MODULE__{opts: lua_opts} = conf) do
    with {:ok, sha} <- get_sha(conf),
         {:ok, key} <- get_key(conn, conf),
         command <- ["EVALSHA", sha, length(key)] ++ key ++ lua_opts,
         {:ok, result} <- apply_command(conf, [command]) do
      {:ok, result}
    else
      {:error, msg} ->
        with true <- noscript_error?(msg),
             {:ok, sha} <- load_sha(conf),
             {:ok, key} <- get_key(conn, conf) do
          command = ["EVALSHA", sha, length(key)] ++ key ++ lua_opts
          apply_command(conf, [command])
        else
          false -> {:error, msg}
          err -> err
        end

      err ->
        err
    end
  end

  defp put_headers(conn, [[header_k, header_v] | headers_t], [_d_h_k | d_h_t]) do
    conn
    |> put_resp_header(header_k, header_v)
    |> put_headers(headers_t, d_h_t)
  end

  defp put_headers(conn, [header_v | headers_t], [d_h_k | d_h_t]) do
    conn
    |> put_resp_header(d_h_k, header_v)
    |> put_headers(headers_t, d_h_t)
  end

  defp put_headers(conn, [[header_k, header_v] | headers_t], []) do
    conn
    |> put_resp_header(header_k, header_v)
    |> put_headers(headers_t, [])
  end

  defp put_headers(conn, [], _default_headers), do: conn

  defp get_sha(%__MODULE__{script_id: script_id} = conf) do
    case :persistent_term.get({__MODULE__, :sha, script_id}, nil) do
      nil -> load_sha(conf)
      sha -> {:ok, sha}
    end
  end

  defp load_sha(%__MODULE__{script: {m, f, a}, script_id: id} = conf) do
    with {:ok, script} <- apply(m, f, a),
         command <- ["SCRIPT", "LOAD", script],
         {:ok, sha} <- apply_command(conf, [command]),
         :ok <- :persistent_term.put({__MODULE__, :sha, id}, sha) do
      {:ok, sha}
    else
      err ->
        :persistent_term.erase(id)
        err
    end
  end

  defp get_key(conn, %__MODULE__{key: {m, f, a}}), do: apply(m, f, [conn, a])

  defp apply_command(%__MODULE__{cmd: {m, f, a}}, command), do: apply(m, f, command ++ a)

  defp apply_response(conn, %__MODULE__{response: {m, f, a}} = conf, eval_result),
    do: apply(m, f, [conn, conf, eval_result, a])

  defp noscript_error?(msg) when is_map(msg) or is_struct(msg),
    do: msg |> Map.get(:message, "") |> noscript_error?()

  defp noscript_error?(msg), do: msg in @no_script_msg
end
